File: python_agent_tools/src/agents/qa_agent.py
"""QA Agent - Testing, validation, and quality assurance."""

from typing import Any, Dict, List, Optional
from .base_agent import BaseAgent, AgentConfig, AgentResult, AgentState
import structlog


class QAAgentConfig(AgentConfig):
    """Configuration for the QA Agent."""
    test_frameworks: List[str] = ["pytest", "unittest", "jest"]
    coverage_threshold: float = 0.8
    max_test_time_seconds: int = 300
    enable_linting: bool = True
    enable_type_checking: bool = True


class QAAgent(BaseAgent):
    """Agent specialized in testing, validation, and quality assurance."""
    
    name = "qa"
    
    def __init__(
        self,
        config: Optional[QAAgentConfig] = None,
        tools: Optional[List] = None,
        memory = None,
    ):
        super().__init__(tools=tools, memory=memory)
        self.config = config or QAAgentConfig(name="qa")
        self._test_results: List[Dict[str, Any]] = []
    
    @property
    def description(self) -> str:
        return "Performs testing, validation, and quality assurance for code and data"
    
    @property
    def category(self) -> str:
        return "quality"
    
    @property
    def tools_used(self) -> List[str]:
        return ["code_execute", "shell", "file_read", "reason"]
    
    def get_system_prompt(self) -> str:
        return """You are the QA Agent, specialized in testing and quality assurance.
Your role is to:
1. Generate and run unit tests for code
2. Perform integration and end-to-end testing
3. Validate data integrity and correctness
4. Run linting and code quality checks
5. Measure and report test coverage
6. Identify bugs and quality issues

Testing capabilities:
- Unit testing (pytest, unittest, jest)
- Integration testing
- API testing
- Data validation
- Performance testing
- Regression testing

Quality checks:
- Code linting (pylint, eslint)
- Type checking (mypy, typescript)
- Security scanning
- Code complexity analysis
- Style guide compliance

Best practices:
- Test edge cases and error conditions
- Maintain high code coverage
- Write clear, maintainable tests
- Document test scenarios
- Report issues with reproduction steps"""
    
    async def run_tests(
        self,
        test_path: str,
        framework: str = "pytest"
    ) -> Dict[str, Any]:
        """Run tests in a directory or file."""
        if framework not in self.config.test_frameworks:
            return {"error": f"Framework '{framework}' not supported"}
        
        command = f"{framework} {test_path}"
        if framework == "pytest":
            command = f"python -m pytest {test_path} -v"
        elif framework == "jest":
            command = f"npx jest {test_path}"
        
        result = await self.execute_tool("shell", {
            "command": command,
            "timeout": self.config.max_test_time_seconds
        })
        
        test_result = {
            "framework": framework,
            "path": test_path,
            "success": result.success,
            "output": result.data if result.success else result.error
        }
        
        self._test_results.append(test_result)
        return test_result
    
    async def generate_tests(self, code: str, language: str = "python") -> str:
        """Generate tests for the given code."""
        result = await self.execute_tool("reason", {
            "task": f"Generate comprehensive unit tests for this {language} code",
            "context": {"code": code, "language": language}
        })
        
        if result.success and result.data is not None and isinstance(result.data, dict):
            return result.data.get("tests", "")
        return ""
    
    async def validate_data(self, data: Any, schema: Dict[str, Any]) -> Dict[str, Any]:
        """Validate data against a schema."""
        result = await self.execute_tool("reason", {
            "task": "Validate this data against the provided schema",
            "context": {"data": str(data)[:5000], "schema": schema}
        })
        
        if result.success and result.data is not None:
            return result.data if isinstance(result.data, dict) else {"result": result.data}
        return {"valid": False, "errors": [result.error]}
    
    async def run_linter(self, path: str, language: str = "python") -> Dict[str, Any]:
        """Run linting on a file or directory."""
        if not self.config.enable_linting:
            return {"error": "Linting disabled"}
        
        command = ""
        if language == "python":
            command = f"python -m pylint {path}"
        elif language in ["javascript", "typescript"]:
            command = f"npx eslint {path}"
        
        result = await self.execute_tool("shell", {"command": command, "timeout": 60})
        
        return {
            "language": language,
            "path": path,
            "issues": result.data if result.success else result.error
        }
    
    async def check_coverage(self, test_path: str) -> Dict[str, Any]:
        """Check test coverage."""
        result = await self.execute_tool("shell", {
            "command": f"python -m pytest {test_path} --cov --cov-report=json",
            "timeout": self.config.max_test_time_seconds
        })
        
        coverage_data = result.data if result.success else {}
        meets_threshold = coverage_data.get("coverage", 0) >= self.config.coverage_threshold
        
        return {
            "coverage": coverage_data,
            "meets_threshold": meets_threshold,
            "threshold": self.config.coverage_threshold
        }
    
    async def type_check(self, path: str) -> Dict[str, Any]:
        """Run type checking on a file or directory."""
        if not self.config.enable_type_checking:
            return {"error": "Type checking disabled"}
        
        result = await self.execute_tool("shell", {
            "command": f"python -m mypy {path}",
            "timeout": 120
        })
        
        return {
            "path": path,
            "issues": result.data if result.success else result.error,
            "success": result.success
        }
    
    async def run(self, task: str, context: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Execute the QA agent's main loop."""
        self.state = AgentState.EXECUTING
        context = context or {}
        
        if "test" in task.lower() and "generate" in task.lower():
            code = context.get("code", "")
            tests = await self.generate_tests(code)
            return {"action": "generate_tests", "tests": tests}
        elif "test" in task.lower():
            path = context.get("path", ".")
            framework = context.get("framework", "pytest")
            result = await self.run_tests(path, framework)
            return {"action": "run_tests", "result": result}
        elif "lint" in task.lower():
            path = context.get("path", ".")
            language = context.get("language", "python")
            result = await self.run_linter(path, language)
            return {"action": "lint", "result": result}
        elif "coverage" in task.lower():
            path = context.get("path", ".")
            result = await self.check_coverage(path)
            return {"action": "coverage", "result": result}
        elif "type" in task.lower():
            path = context.get("path", ".")
            result = await self.type_check(path)
            return {"action": "type_check", "result": result}
        elif "validate" in task.lower():
            data = context.get("data")
            schema = context.get("schema", {})
            result = await self.validate_data(data, schema)
            return {"action": "validate", "result": result}
        else:
            path = context.get("path", ".")
            result = await self.run_tests(path)
            return {"action": "run_tests", "result": result}
    
    async def plan(self, task: str, context: Dict[str, Any]) -> List[str]:
        """Generate a plan for the QA task."""
        return [f"Execute QA task: {task}"]
    
    async def execute(self, task: str, context: Optional[Dict[str, Any]] = None) -> AgentResult:
        """Execute a QA task."""
        self.logger.info("qa_execute", task=task[:100] if task else "")
        self.state = AgentState.EXECUTING
        
        try:
            result = await self.run(task, context)
            self.state = AgentState.COMPLETED
            return AgentResult(
                success=True,
                data=result,
                metadata={"tests_run": len(self._test_results)}
            )
        except Exception as e:
            self.state = AgentState.ERROR
            self.logger.error("qa_error", error=str(e))
            return AgentResult(success=False, error=str(e))
    
    async def initialize(self) -> None:
        """Initialize the QA agent."""
        await super().initialize()
        self._test_results = []
        self.logger.info("qa_agent_initialized", frameworks=self.config.test_frameworks)
    
    async def shutdown(self) -> None:
        """Shutdown the QA agent."""
        self._test_results.clear()
        await super().shutdown()
        self.logger.info("qa_agent_shutdown")

File: server/agent/langgraph/agents/QAAgent.ts
import { z } from "zod";
import OpenAI from "openai";
import { BaseAgent, BaseAgentConfig, AgentTask, AgentResult, AgentCapability } from "./types";

const xaiClient = new OpenAI({
  baseURL: "https://api.x.ai/v1",
  apiKey: process.env.XAI_API_KEY || "missing",
});

const DEFAULT_MODEL = "grok-4-1-fast-non-reasoning";

export class QAAgent extends BaseAgent {
  constructor() {
    const config: BaseAgentConfig = {
      name: "QAAgent",
      description: "Specialized agent for quality assurance, testing, validation, and verification. Expert at identifying issues and ensuring quality standards.",
      model: DEFAULT_MODEL,
      temperature: 0.1,
      maxTokens: 8192,
      systemPrompt: `You are the QAAgent - an expert quality assurance engineer.

Your capabilities:
1. Test Case Generation: Create comprehensive test cases from requirements
2. Test Execution: Plan and execute testing strategies
3. Bug Identification: Find defects and edge cases
4. Validation: Verify outputs against specifications
5. Performance Testing: Identify bottlenecks and issues
6. Accessibility Testing: Ensure WCAG compliance

Testing methodology:
- Requirements-based testing
- Boundary value analysis
- Equivalence partitioning
- Error guessing
- Exploratory testing
- Regression testing

Quality metrics:
- Code coverage
- Defect density
- Test pass rate
- Performance benchmarks
- Accessibility scores

Output formats:
- Test cases in structured format
- Bug reports with reproduction steps
- Test execution reports
- Quality dashboards`,
      tools: ["code_test", "code_review", "verify", "health_check"],
      timeout: 180000,
      maxIterations: 25,
    };
    super(config);
  }

  async execute(task: AgentTask): Promise<AgentResult> {
    const startTime = Date.now();
    this.updateState({ status: "running", currentTask: task.description, startedAt: new Date().toISOString() });

    try {
      const qaTaskType = this.determineQATaskType(task);
      let result: any;

      switch (qaTaskType) {
        case "generate_tests":
          result = await this.generateTestCases(task);
          break;
        case "validate":
          result = await this.validateOutput(task);
          break;
        case "bug_hunt":
          result = await this.huntBugs(task);
          break;
        case "performance":
          result = await this.performanceTest(task);
          break;
        case "accessibility":
          result = await this.accessibilityTest(task);
          break;
        default:
          result = await this.handleGeneralQA(task);
      }

      this.updateState({ status: "completed", progress: 100, completedAt: new Date().toISOString() });

      return {
        taskId: task.id,
        agentId: this.state.id,
        success: true,
        output: result,
        duration: Date.now() - startTime,
      };
    } catch (error: any) {
      this.updateState({ status: "failed", error: error.message });
      return {
        taskId: task.id,
        agentId: this.state.id,
        success: false,
        error: error.message,
        duration: Date.now() - startTime,
      };
    }
  }

  private determineQATaskType(task: AgentTask): string {
    const description = task.description.toLowerCase();
    if (description.includes("test case") || description.includes("generate test")) return "generate_tests";
    if (description.includes("validate") || description.includes("verify")) return "validate";
    if (description.includes("bug") || description.includes("defect") || description.includes("issue")) return "bug_hunt";
    if (description.includes("performance") || description.includes("load") || description.includes("stress")) return "performance";
    if (description.includes("accessibility") || description.includes("a11y") || description.includes("wcag")) return "accessibility";
    return "general";
  }

  private async generateTestCases(task: AgentTask): Promise<any> {
    const requirements = task.input.requirements || task.description;
    const testType = task.input.testType || "functional";

    const messages: any[] = [
      { role: "system", content: this.config.systemPrompt },
      {
        role: "user",
        content: `Generate ${testType} test cases for:\n${requirements}\n\nAdditional context: ${JSON.stringify(task.input)}\n\nReturn JSON:\n{\n  "testSuite": {\n    "name": "test suite name",\n    "description": "what is being tested",\n    "testCases": [\n      {\n        "id": "TC001",\n        "name": "test case name",\n        "description": "what is tested",\n        "preconditions": ["setup required"],\n        "steps": [{"step": 1, "action": "", "expectedResult": ""}],\n        "priority": "high|medium|low",\n        "type": "positive|negative|edge"\n      }\n    ]\n  },\n  "coverage": {\n    "requirements": ["covered requirements"],\n    "gaps": ["areas not covered"]\n  },\n  "estimatedDuration": "time to execute all tests"\n}`,
      },
    ];

    let content = "";
    let isComplete = false;
    let iteration = 0;

    while (!isComplete && iteration < this.config.maxIterations) {
      iteration++;
      const response = await xaiClient.chat.completions.create({
        model: this.config.model,
        messages,
        temperature: 0.1,
        tools: [
          {
            type: "function",
            function: {
              name: "delegate_task",
              description: "Delegate a sub-task to a specialized agent.",
              parameters: {
                type: "object",
                properties: {
                  targetAgentName: { type: "string" },
                  subTaskDescription: { type: "string" },
                  inputData: { type: "object" }
                },
                required: ["targetAgentName", "subTaskDescription"]
              }
            }
          }
        ],
        tool_choice: "auto",
      });

      const message = response.choices[0].message;
      messages.push(message);

      if (message.tool_calls) {
        for (const toolCall of message.tool_calls) {
          if (toolCall.type === "function" && toolCall.function?.name === "delegate_task") {
            const args = JSON.parse(toolCall.function.arguments);
            try {
              const result = await this.delegateTask(args.targetAgentName, args.subTaskDescription, args.inputData || {});
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: JSON.stringify(result.output)
              });
            } catch (err: any) {
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: `Error delegating task: ${err.message}`
              });
            }
          }
        }
      } else {
        content = message.content || "{}";
        isComplete = true;
      }
    }

    const jsonMatch = content.match(/\{[\s\S]*\}/);

    return {
      type: "test_generation",
      testCases: jsonMatch ? JSON.parse(jsonMatch[0]) : { description: content },
      timestamp: new Date().toISOString(),
    };
  }

  private async validateOutput(task: AgentTask): Promise<any> {
    const actual = task.input.actual || "";
    const expected = task.input.expected || "";
    const criteria = task.input.criteria || [];

    const messages: any[] = [
      { role: "system", content: this.config.systemPrompt },
      {
        role: "user",
        content: `Validate output against specification:\nActual: ${JSON.stringify(actual)}\nExpected: ${JSON.stringify(expected)}\nCriteria: ${JSON.stringify(criteria)}\nTask: ${task.description}\n\nReturn JSON:\n{\n  "valid": boolean,\n  "score": 0-100,\n  "checks": [\n    {"criterion": "", "passed": boolean, "details": "", "severity": "low|medium|high"}\n  ],\n  "discrepancies": ["list of differences"],\n  "recommendations": ["suggestions for improvement"]\n}`,
      },
    ];

    let content = "";
    let isComplete = false;
    let iteration = 0;

    while (!isComplete && iteration < this.config.maxIterations) {
      iteration++;
      const response = await xaiClient.chat.completions.create({
        model: this.config.model,
        messages,
        temperature: 0.1,
        tools: [
          {
            type: "function",
            function: {
              name: "delegate_task",
              description: "Delegate a sub-task to a specialized agent.",
              parameters: {
                type: "object",
                properties: {
                  targetAgentName: { type: "string" },
                  subTaskDescription: { type: "string" },
                  inputData: { type: "object" }
                },
                required: ["targetAgentName", "subTaskDescription"]
              }
            }
          }
        ],
        tool_choice: "auto",
      });

      const message = response.choices[0].message;
      messages.push(message);

      if (message.tool_calls) {
        for (const toolCall of message.tool_calls) {
          if (toolCall.type === "function" && toolCall.function?.name === "delegate_task") {
            const args = JSON.parse(toolCall.function.arguments);
            try {
              const result = await this.delegateTask(args.targetAgentName, args.subTaskDescription, args.inputData || {});
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: JSON.stringify(result.output)
              });
            } catch (err: any) {
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: `Error delegating task: ${err.message}`
              });
            }
          }
        }
      } else {
        content = message.content || "{}";
        isComplete = true;
      }
    }

    const jsonMatch = content.match(/\{[\s\S]*\}/);

    return {
      type: "validation",
      validation: jsonMatch ? JSON.parse(jsonMatch[0]) : { description: content },
      timestamp: new Date().toISOString(),
    };
  }

  private async huntBugs(task: AgentTask): Promise<any> {
    const code = task.input.code || "";
    const context = task.input.context || "";

    const messages: any[] = [
      { role: "system", content: this.config.systemPrompt },
      {
        role: "user",
        content: `Find bugs and issues in:\n\`\`\`\n${code}\n\`\`\`\n\nContext: ${context}\nTask: ${task.description}\n\nReturn JSON:\n{\n  "bugs": [\n    {\n      "id": "BUG001",\n      "severity": "critical|high|medium|low",\n      "type": "logic|security|performance|ux",\n      "location": "file:line",\n      "description": "what's wrong",\n      "reproductionSteps": ["how to reproduce"],\n      "suggestedFix": "how to fix",\n      "confidence": 0-100\n    }\n  ],\n  "codeSmells": ["potential issues"],\n  "edgeCases": ["unhandled scenarios"],\n  "overallRisk": "low|medium|high"\n}`,
      },
    ];

    let content = "";
    let isComplete = false;
    let iteration = 0;

    while (!isComplete && iteration < this.config.maxIterations) {
      iteration++;
      const response = await xaiClient.chat.completions.create({
        model: this.config.model,
        messages,
        temperature: 0.1,
        tools: [
          {
            type: "function",
            function: {
              name: "delegate_task",
              description: "Delegate a sub-task to a specialized agent.",
              parameters: {
                type: "object",
                properties: {
                  targetAgentName: { type: "string" },
                  subTaskDescription: { type: "string" },
                  inputData: { type: "object" }
                },
                required: ["targetAgentName", "subTaskDescription"]
              }
            }
          }
        ],
        tool_choice: "auto",
      });

      const message = response.choices[0].message;
      messages.push(message);

      if (message.tool_calls) {
        for (const toolCall of message.tool_calls) {
          if (toolCall.type === "function" && toolCall.function?.name === "delegate_task") {
            const args = JSON.parse(toolCall.function.arguments);
            try {
              const result = await this.delegateTask(args.targetAgentName, args.subTaskDescription, args.inputData || {});
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: JSON.stringify(result.output)
              });
            } catch (err: any) {
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: `Error delegating task: ${err.message}`
              });
            }
          }
        }
      } else {
        content = message.content || "{}";
        isComplete = true;
      }
    }

    const jsonMatch = content.match(/\{[\s\S]*\}/);

    return {
      type: "bug_hunting",
      findings: jsonMatch ? JSON.parse(jsonMatch[0]) : { description: content },
      timestamp: new Date().toISOString(),
    };
  }

  private async performanceTest(task: AgentTask): Promise<any> {
    const target = task.input.target || "";
    const metrics = task.input.metrics || ["response_time", "throughput", "memory"];

    const messages: any[] = [
      { role: "system", content: this.config.systemPrompt },
      {
        role: "user",
        content: `Create performance test plan for:\nTarget: ${target}\nMetrics: ${JSON.stringify(metrics)}\nTask: ${task.description}\n\nReturn JSON:\n{\n  "testPlan": {\n    "scenarios": [\n      {\n        "name": "scenario name",\n        "load": "concurrent users/requests",\n        "duration": "test duration",\n        "rampUp": "ramp up time"\n      }\n    ],\n    "metrics": ["metrics to measure"],\n    "thresholds": {"response_time": "< 200ms", "error_rate": "< 1%"}\n  },\n  "tools": ["recommended tools"],\n  "code": {\n    "k6": "k6 script",\n    "artillery": "artillery config"\n  }\n}`,
      },
    ];

    let content = "";
    let isComplete = false;
    let iteration = 0;

    while (!isComplete && iteration < this.config.maxIterations) {
      iteration++;
      const response = await xaiClient.chat.completions.create({
        model: this.config.model,
        messages,
        temperature: 0.1,
        tools: [
          {
            type: "function",
            function: {
              name: "delegate_task",
              description: "Delegate a sub-task to a specialized agent.",
              parameters: {
                type: "object",
                properties: {
                  targetAgentName: { type: "string" },
                  subTaskDescription: { type: "string" },
                  inputData: { type: "object" }
                },
                required: ["targetAgentName", "subTaskDescription"]
              }
            }
          }
        ],
        tool_choice: "auto",
      });

      const message = response.choices[0].message;
      messages.push(message);

      if (message.tool_calls) {
        for (const toolCall of message.tool_calls) {
          if (toolCall.type === "function" && toolCall.function?.name === "delegate_task") {
            const args = JSON.parse(toolCall.function.arguments);
            try {
              const result = await this.delegateTask(args.targetAgentName, args.subTaskDescription, args.inputData || {});
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: JSON.stringify(result.output)
              });
            } catch (err: any) {
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: `Error delegating task: ${err.message}`
              });
            }
          }
        }
      } else {
        content = message.content || "{}";
        isComplete = true;
      }
    }

    const jsonMatch = content.match(/\{[\s\S]*\}/);

    return {
      type: "performance_testing",
      testPlan: jsonMatch ? JSON.parse(jsonMatch[0]) : { description: content },
      timestamp: new Date().toISOString(),
    };
  }

  private async accessibilityTest(task: AgentTask): Promise<any> {
    const target = task.input.target || task.description;
    const standard = task.input.standard || "WCAG 2.1 AA";

    const messages: any[] = [
      { role: "system", content: this.config.systemPrompt },
      {
        role: "user",
        content: `Create accessibility test checklist for:\nTarget: ${target}\nStandard: ${standard}\nTask: ${task.description}\n\nReturn JSON:\n{\n  "checklist": [\n    {\n      "criterion": "WCAG criterion",\n      "level": "A|AA|AAA",\n      "category": "perceivable|operable|understandable|robust",\n      "testMethod": "how to test",\n      "automatable": boolean\n    }\n  ],\n  "tools": ["recommended a11y tools"],\n  "commonIssues": ["likely accessibility issues to check"]\n}`,
      },
    ];

    let content = "";
    let isComplete = false;
    let iteration = 0;

    while (!isComplete && iteration < this.config.maxIterations) {
      iteration++;
      const response = await xaiClient.chat.completions.create({
        model: this.config.model,
        messages,
        temperature: 0.1,
        tools: [
          {
            type: "function",
            function: {
              name: "delegate_task",
              description: "Delegate a sub-task to a specialized agent.",
              parameters: {
                type: "object",
                properties: {
                  targetAgentName: { type: "string" },
                  subTaskDescription: { type: "string" },
                  inputData: { type: "object" }
                },
                required: ["targetAgentName", "subTaskDescription"]
              }
            }
          }
        ],
        tool_choice: "auto",
      });

      const message = response.choices[0].message;
      messages.push(message);

      if (message.tool_calls) {
        for (const toolCall of message.tool_calls) {
          if (toolCall.type === "function" && toolCall.function?.name === "delegate_task") {
            const args = JSON.parse(toolCall.function.arguments);
            try {
              const result = await this.delegateTask(args.targetAgentName, args.subTaskDescription, args.inputData || {});
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: JSON.stringify(result.output)
              });
            } catch (err: any) {
              messages.push({
                role: "tool",
                tool_call_id: toolCall.id,
                content: `Error delegating task: ${err.message}`
              });
            }
          }
        }
      } else {
        content = message.content || "{}";
        isComplete = true;
      }
    }

    const jsonMatch = content.match(/\{[\s\S]*\}/);

    return {
      type: "accessibility_testing",
      checklist: jsonMatch ? JSON.parse(jsonMatch[0]) : { description: content },
      timestamp: new Date().toISOString(),
    };
  }

  private async handleGeneralQA(task: AgentTask): Promise<any> {
    const response = await xaiClient.chat.completions.create({
      model: this.config.model,
      messages: [
        { role: "system", content: this.config.systemPrompt },
        { role: "user", content: `QA task: ${task.description}\nInput: ${JSON.stringify(task.input)}` },
      ],
      temperature: 0.1,
    });

    return {
      type: "general_qa",
      result: response.choices[0].message.content,
      timestamp: new Date().toISOString(),
    };
  }

  getCapabilities(): AgentCapability[] {
    return [
      {
        name: "generate_tests",
        description: "Generate test cases from requirements",
        inputSchema: z.object({ requirements: z.string(), testType: z.string().optional() }),
        outputSchema: z.object({ testCases: z.array(z.any()), coverage: z.any() }),
      },
      {
        name: "validate",
        description: "Validate output against specifications",
        inputSchema: z.object({ actual: z.any(), expected: z.any() }),
        outputSchema: z.object({ valid: z.boolean(), discrepancies: z.array(z.string()) }),
      },
      {
        name: "find_bugs",
        description: "Identify bugs and issues in code",
        inputSchema: z.object({ code: z.string() }),
        outputSchema: z.object({ bugs: z.array(z.any()) }),
      },
    ];
  }
}

export const qaAgent = new QAAgent();
